libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java

```diff
@@ -1047,6 +1049,33 @@ public XContentBuilder copyCurrentStructure(XContentParser parser) throws IOExce
         return this;
     }
 
+    /**
+     * Used in {@link #writableFieldAsBase64(String, WritableValue)}
+     */
+    @FunctionalInterface
+    public interface WritableValue {
+        void writeTo(OutputStream os) throws IOException;
+    }
+
+    /**
+     * Write the serialization of a {@link WritableValue} via {@link WritableValue#writeTo(OutputStream)} as a string encoded in
+     * Base64 format. This API can be used to generate XContent directly without the intermediate results to reduce memory usage.
+     */
+    public XContentBuilder writableFieldAsBase64(String name, WritableValue value) throws IOException {
+        generator.writeFieldDirectly(name, os -> {
+            os.write('\"');
+            try (OutputStream encoded = Base64.getEncoder().wrap(new FilterOutputStream(os) {
+                @Override
+                public void close() {
```

Comment:
I wonder if it's be clearer not to use the try catch and have a comment about how you intentionally army closing. Not sure!

Reply:
I added a comment. We need to close the output stream that is wrapped the Base64 encoder to flush the outstanding buffer but we can't close the underlying output stream of the XContentBuilder.

Reply:
Thanks! That makes perfect sense.